src/include/cart/api.h

```diff
@@ -1532,6 +1532,44 @@ typedef void
 int
 crt_register_eviction_cb(crt_eviction_cb cb, void *arg);
 
+enum crt_event_source {
+	CRT_EVS_UNKNOWN,
+	CRT_EVS_PMIX,
+	CRT_EVS_SWIM,
+};
+
+enum crt_event_type {
+	CRT_EVT_ALIVE,
+	CRT_EVT_DEAD,
+};
+
+typedef void
+(*crt_event_cb) (d_rank_t rank, enum crt_event_source src,
+		 enum crt_event_type type, void *arg);
+
+/**
+ * This function registers an event handler for any changes in rank state.
+ * The event should be processed in none blocking mode because this
+ * handler is called under lock which should not be held for long time.
+ * Sleeping is also prohibited in this handler! Only quick reaction is
+ * expected into this handler before return.
+ *
+ * \param[in] event_handler    event handler to register
+ * \param[in] arg              arg to event_handler
+ *
+ * \return                     DER_SUCCESS on success, negative value on error
+ */
+int  crt_register_event_cb(crt_event_cb event_handler, void *arg);
+
+/**
+ * Unregister an event handler. The pair of arguments (event_handler and arg)
+ * should be same as the ones provided during registration.
+ *
+ * \param[in] event_handler    event handler to register
+ * \param[in] arg              arg to event_handler
+ */
+void crt_unregister_event_cb(crt_event_cb event_handler, void *arg);
```

Comment:
crt_unregister_progress_cb() returns an int, what is the justification for this function not?  I think we should be consistent between the two here.